### ISSUE TITLE
Custom plugins using `register_transform` don't seem to work for dynamic attributes

### ISSUE THREAD

**Issue by zapstar** [PyCQA/pylint#2878]
### Steps to reproduce

1. Create a file with a class that dynamically populates some attributes called `hello.py`. My plugin's purpose is to use some astroid logic and identify the attributes for all classes correctly.

```python
class Hello:
    def __init__(self):
        # self.x = 1
        # self.y = 2
        # Set two above attributes dynamically
        d = {'x': 1, 'y': 2}
        for k, v in d.items():
            setattr(self, k, v)
```

2. Create a file that uses the module `hello` called `use_hello.py`. This file will be checked by `PyLint`.

```python
"""
Use `hello` module
"""
import hello


def main():
    """Runner function"""
    var = hello.Hello()
    print(var.x)
    print(var.y)

if __name__ == '__main__':
    main()
```

3. Create a plugin module to be used by `PyLint` called `register_transform.py`. The purpose of this plugin is to transform all classes inside this module and return a new module with the correctly populated dynamic attributes. This is a minimalistic example with only static data, but I'll want to generate code similar to the class `Hello` depending on the class names & attribute names inside the module.

```python
import astroid

def register(_):
    pass

def _transform(_):
    return astroid.parse("""
class Hello(object):
    def __init__(self):
        self.x = 1
        self.y = 2
""")

def _looks_like_hello(node):
    return node.qname() == "hello"

astroid.MANAGER.register_transform(astroid.Module, _transform, _looks_like_hello)
```

### Current behavior
```
$ pylint --init-hook='import sys; sys.path.insert(0, ".")' --load-plugins register_transform use_hello.py
************* Module use_hello
use_hello.py:10:10: E1101: Instance of 'Hello' has no 'x' member (no-member)
use_hello.py:11:10: E1101: Instance of 'Hello' has no 'y' member (no-member)

--------------------------------------------------------------------
Your code has been rated at -4.29/10 (previous run: -4.29/10, +0.00)

```

### Expected behavior
```
$ pylint --init-hook='import sys; sys.path.insert(0, ".")' --load-plugins register_transform use_hello.py

---------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: -4.29/10, +14.29)

```

I would have been happy with using `register_module_extender` (which works by the way). But, I'll need to inspect the module nodes and create a new module node with the same class names. This is not possible via `register_module_extender`.

### pylint --version output
```
$ pylint --version
pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, Mar 27 2019, 09:23:15) 
[Clang 10.0.1 (clang-1001.0.46.3)]
```

**Comment by PCManticore** [comment]
Hi @zapstar Thanks for reporting this issue.

I can confirm there is a bug here, but it might be tricky to solve. This seems to happen only for transforms that replace node modules wholesale, such as your transform. While the transform works and replaces the node in the AST, when the `no-member` check gets triggered, the inference bypasses the transformed module, probably due to a miscommunication between `astroid`'s import system and the the transforms themselves, importing instead the non transformed module.

Ideally this should work, but in the meantime you can either use `register_module_extender` or modify your transform so instead of returning a complete module, it modifies the current node in place, as in:
```

def _transform(node):
    hello = astroid.parse("""
class Hello(object):
    def __init__(self):
        self.x = 1
        self.y = 2
""").body[0]
    node.locals['Hello'] = [hello]
    return node
```

**Comment by zapstar** [comment]
I don't think it works for other types of nodes either. Please see my example. Or maybe it works for only non-scoped nodes?

Third party module: `hello.py`
```python
class Dynamic:
    def __init__(self):
        # Set two attributes
        # x to 1
        # y to 2
        d = {'x': 1, 'y': 2}
        for k, v in d.items():
            setattr(self, k, v)

class Static:
    def __init__(self):
        self.x = 1
        self.y = 2
```

File `use_classes.py` using `hello.py`
```python
"""
Use `hello` module
"""
import hello


def main():
    """Runner function"""
    # Dynamic class
    var1 = hello.Dynamic()
    print(var1.x)
    print(var1.y)
    print(var1.z)
    var1.w = 20
    # Static class
    var2 = hello.Static()
    print(var2.x)
    print(var2.y)
    print(var2.z)
    var2.w = 20

if __name__ == '__main__':
    main()

```

Plugin `register_transform_class.py`

```python
import astroid

def register(_):
    pass

def _transform(node):
    module = astroid.parse("""
class Dynamic(object):
    def __init__(self):
        self.x = 1
        self.y = 2
""")
    new_node = module.locals['Dynamic'][0]
    print(new_node)
    return new_node

def _looks_like_hello(node):
    return node.qname() == "hello.Dynamic"

astroid.MANAGER.register_transform(astroid.ClassDef, _transform, _looks_like_hello)
```

Expected output
```
$ pylint --init-hook='import sys; sys.path.insert(0, ".")' --load-plugins register_transform_class use_classes.py
ClassDef.Dynamic(name='Dynamic',
                 doc=None,
                 decorators=None,
                 bases=[<Name.object l.2 at 0x1045ba668>],
                 body=[<FunctionDef.__init__ l.3 at 0x1045ba6a0>])
************* Module use_classes
use_classes.py:13:10: E1101: Instance of 'Dynamic' has no 'z' member; maybe 'x'? (no-member)
use_classes.py:19:10: E1101: Instance of 'Static' has no 'z' member; maybe 'x'? (no-member)

--------------------------------------------------------------------
Your code has been rated at -4.29/10 (previous run: -4.29/10, +0.00)

```

Current output
```
$ pylint --init-hook='import sys; sys.path.insert(0, ".")' --load-plugins register_transform_class use_classes.py
ClassDef.Dynamic(name='Dynamic',
                 doc=None,
                 decorators=None,
                 bases=[<Name.object l.2 at 0x1045ba668>],
                 body=[<FunctionDef.__init__ l.3 at 0x1045ba6a0>])
************* Module use_classes
use_classes.py:11:10: E1101: Instance of 'Dynamic' has no 'x' member; maybe 'w'? (no-member)
use_classes.py:12:10: E1101: Instance of 'Dynamic' has no 'y' member; maybe 'w'? (no-member)
use_classes.py:13:10: E1101: Instance of 'Dynamic' has no 'z' member; maybe 'w'? (no-member)
use_classes.py:19:10: E1101: Instance of 'Static' has no 'z' member; maybe 'x'? (no-member)

--------------------------------------------------------------------
Your code has been rated at -4.29/10 (previous run: -4.29/10, +0.00)

```

**Comment by PCManticore** [comment]
Thanks @zapstar Adjusted the title to make it obvious that it does not work for other nodes as well.

**Comment by jacobtylerwalls** [comment]
Thanks both, as far as I see it based on reading this for the first time, this issue is the root cause for why this will emit a `no-member` warning:

```
class A:
    def __init__(self):
        setattr('x', None)

    def getX(self):
        return self.x
```
[E1101(no-member), A.getX] Instance of 'A' has no 'x' member
